src/Data/Text/Rope.hs

```diff
@@ -129,57 +172,55 @@ length = \case
 -- Position {posLine = 2, posColumn = 0}
 --
 lengthAsPosition :: Rope -> Position
-lengthAsPosition = \case
-  Empty -> mempty
-  Node _ _ _ _ p -> p
+lengthAsPosition rp =
+  Position nls (length line)
+  where
+    nls = newlines rp
+    (_, line) = splitAtLine nls rp
 
 instance Semigroup Rope where
   Empty <> t = t
   t <> Empty = t
-  Node l1 c1 r1 u1 p1 <> Node l2 c2 r2 u2 p2 = defragment
+  Node l1 c1 r1 m1 <> Node l2 c2 r2 m2 = defragment
     l1
     c1
-    (Node (r1 <> l2) c2 r2 (length r1 + u2) (lengthAsPosition r1 <> p2))
-    (u1 + u2)
-    (p1 <> p2)
+    (Node (r1 <> l2) c2 r2 (metrics r1 <> m2))
+    (m1 <> m2)
 
 instance Monoid Rope where
   mempty = Empty
   mappend = (<>)
 
-defragment :: HasCallStack => Rope -> TL.TextLines -> Rope -> Word -> Position -> Rope
-defragment !l !c !r !u !p
+defragment :: HasCallStack => Rope -> TL.TextLines -> Rope -> Metrics -> Rope
+defragment !l !c !r !m
 #ifdef DEBUG
   | TL.null c = error "Data.Text.Lines: violated internal invariant"
 #endif
-  | u < DEFRAGMENTATION_THRESHOLD
-  = Node Empty (toTextLines rp) Empty u p
+  | _metricsCharLen m < DEFRAGMENTATION_THRESHOLD
+  = Node Empty (toTextLines rp) Empty m
   | otherwise
   = rp
   where
-    rp = Node l c r u p
+    rp = Node l c r m
 
 -- | Create from 'TL.TextLines', linear time.
 fromTextLines :: TL.TextLines -> Rope
 fromTextLines tl
   | TL.null tl = Empty
-  | otherwise = Node Empty tl Empty (TL.length tl) (TL.lengthAsPosition tl)
+  | otherwise = Node Empty tl Empty (linesMetrics tl)
 
-node :: HasCallStack => Rope -> TL.TextLines -> Rope -> Rope
-node l c r = defragment l c r totalLength totalLengthAsPosition
-  where
-    totalLength = length l + TL.length c + length r
-    totalLengthAsPosition = lengthAsPosition l <> TL.lengthAsPosition c <> lengthAsPosition r
+node :: HasCallStack => Rope -> TL.TextLines -> Metrics -> Rope -> Rope
+node l c cm r = defragment l c r (metrics l <> cm <> metrics r)
 
-(|>) :: Rope -> TL.TextLines -> Rope
-tr |> tl
+snoc :: Rope -> TL.TextLines -> Metrics -> Rope
```

Comment:
Please add a comment explaining the invariant connecting `TL.TextLines` and `Metrics`.